trl/trainer/accelerate_ppo.py

```diff
@@ -0,0 +1,542 @@
+        self,
+        model: Optional[nn.Module]=None, 
+        ref_model: Optional[nn.Module]=None, 
+        tokenizer: Optional[PreTrainedTokenizer]=None, 
+        **config
+    ):
+        """
+        Initialize PPOTrainer.
+        Args:
+            model (torch.model): Hugging Face transformer GPT2 model with value head
+            ref_model (torch.model): Hugging Face transformer GPT2 refrence model used for KL penalty
+            tokenizer (tokenizer): Hugging Face tokenizer
+            ppo_params (dict or None): PPO parameters for training. Can include following keys:
+                'lr' (float): Adam learning rate, default: 1.41e-5
+                'batch_size' (int): Number of samples per optimisation step, default: 256
+                'forward_batch_size' (int): Number of samples forward passed through model at a time, default: 16
+                'ppo_epochs' (int): Number of optimisation epochs per batch of samples, default: 4
+                'gamma' (float)): Gamma parameter for advantage calculation, default: 1.
+                'lam' (float): Lambda parameter for advantage calcualation, default: 0.95
+                'cliprange_value' (float): Range for clipping values in loss calculation, default: 0.2
+                'cliprange' (float): Range for clipping in PPO policy gradient loss, default: 0.2
+                'vf_coef' (float): Scaling factor for value loss, default: 0.1
+                'adap_kl_ctrl' (bool): Use adaptive KL control, otherwise linear, default: True
+                'init_kl_coef' (float): Initial KL penalty coefficient (used for adaptive and linear control), default: 0.2
+                'target' (float): Target KL value for adaptive KL control, default: 6.0
+                'horizon' (float): Horizon for adaptive KL control, default: 10000
+        """
+        super().__init__(self.default_params)
+
+        # Step 1: Initialize Accelerator
+        self.accelerator = Accelerator(log_with="wandb")
+        self.config.update(config)
+
+        # Step 2: Initialize model, tokenizer and dataset        
+        self._build_models_and_tokenizer(model, ref_model, tokenizer)
+        self._build_dataset()
+
+
+        # Step 3: Initialize optimizer and data collator        
+        self.data_collator = DataCollatorForLanguageModeling(self.tokenizer, mlm=False)
+        self.optimizer = Adam(self.model.parameters(), lr=self.config['lr'])
+
+        if self.config['adap_kl_ctrl']:
+            self.kl_ctl = AdaptiveKLController(self.config['init_kl_coef'],
+                                               self.config['target'],
+                                               self.config['horizon'])
+        else:
+            self.kl_ctl = FixedKLController(self.config['init_kl_coef'])
+
+        self.model, self.ref_model, self.optimizer, self.data_collator, self.dataloader = self.accelerator.prepare(self.model, self.ref_model, self.optimizer, self.data_collator, self.dataloader)
```

Comment:
In general we won't have a dataloader here, right?

Reply:
Actually we should, since it is created by `_build_dataset`. But this should be provided by the user

Reply:
What's the use of the dataset inside the trainer? The user will need to get the queries and generate responses right?

Reply:
I think the main use of the dataloader inside the trained is to be prepared correctly by the `accelerator` object -> in case of Mutli-GPU (DP) setup it'll be converted to the relevant wrapper/object

Reply:
I think this workflow can be def optimized

Reply:
Should be fixed in [7516b37](https://github.com/lvwerra/trl/pull/58/commits/7516b37181ded8311bfaad33802829e9767e3487)

Reply:
(also discussed offline about DP)